test/extensions/implementation/autograd.py

```diff
@@ -82,12 +80,33 @@ def extract_ith_element_of_diag_ggn(i, p):
             diag_ggn_p = torch.zeros_like(p).view(-1)
 
             for parameter_index in range(p.numel()):
-                diag_value = extract_ith_element_of_diag_ggn(parameter_index, p)
+                diag_value = extract_ith_element_of_diag_ggn(
+                    parameter_index, p, loss, output
+                )
                 diag_ggn_p[parameter_index] = diag_value
 
             diag_ggns.append(diag_ggn_p.view(p.size()))
         return diag_ggns
 
+    def diag_ggn(self):
+        _, output, loss = self.problem.forward_pass()
+        return self._get_diag_ggn(loss, output)
+
+    def diag_ggn_batch(self):
+        batch_size = self.problem.input.shape[0]
+        _, _, batch_loss = self.problem.forward_pass()
+        loss_list = torch.zeros(batch_size, device=self.problem.device)
+
+        batch_diag_ggn = []
+        for b in range(batch_size):
+            _, output, loss = self.problem.forward_pass(sample_idx=b)
+            diag_ggn = self._get_diag_ggn(loss, output)
+            batch_diag_ggn.append(diag_ggn)
+            loss_list[b] = loss
+        factor = self.problem.get_reduction_factor(batch_loss, loss_list)
+        params_batch_diag_ggn = list(zip(*batch_diag_ggn))
+        return [torch.stack(param) * factor for param in params_batch_diag_ggn]
```

Comment:
Can you add a comment on the data layout?

Something like:

`batch_diag_ggn` has entries `[sample_idx][param_idx]`
`params_batch_diag_ggn` has entries `[param_idx][sample_idx]`

Reply:
I quite don't understand this comment, I have added exactly the same. But would it be better with shapes?